crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -340,6 +340,80 @@ def services
       end
       handle_asynchronously :services
 
+      def openstackbackup
+        ::Crowbar::UpgradeStatus.new.start_step(:backup_openstack)
+
+        dump_path = "/var/lib/crowbar/upgrade/6-to-7-openstack_dump.sql"
+        if File.exist?(dump_path)
+          Rails.logger.debug("OpenStack backup already exists. Skipping...")
+          return
+        end
+
+        psql = postgres_params
+        query = "SELECT SUM(pg_database_size(pg_database.datname)) FROM pg_database;"
+        cmd = "PGPASSWORD=#{psql[:pass]} psql -t -h #{psql[:host]} -U #{psql[:user]} -c '#{query}'"
+
+        Rails.logger.debug("Checking size of OpenStack database")
+        db_size = run_cmd(cmd)
+        unless db_size[:exit_code].zero?
+          Rails.logger.error(
+            "Failed to check size of OpenStack database: #{db_size[:stdout_and_stderr]}"
+          )
+          raise ::Crowbar::Error::UpgradeDatabaseSizeError.new(
+            db_size[:stdout_and_stderr]
+          )
+        end
+
+        free_space = run_cmd(
+          "LANG=C df -x 'tmpfs' -x 'devtmpfs' -B1 -l --output='avail' / | tail -n1"
+        )
+        unless free_space[:exit_code].zero?
+          Rails.logger.error("Cannot determine free disk space: #{free_space[:stdout_and_stderr]}")
+          raise ::Crowbar::Error::UpgradeFreeDiskSpaceError.new(
+            free_space[:stdout_and_stderr]
+          )
+        end
+        if free_space[:stdout_and_stderr].strip.to_i < db_size[:stdout_and_stderr].strip.to_i
+          Rails.logger.error("Not enough free disk space to create the OpenStack database dump")
+          raise ::Crowbar::Error::UpgradeNotEnoughDiskSpaceError.new
+        end
+
+        Rails.logger.debug("Creating OpenStack database dump")
+        db_dump = run_cmd(
+          "PGPASSWORD=#{psql[:pass]} pg_dumpall -h #{psql[:host]} -U #{psql[:user]} > #{dump_path}"
+        )
+        unless db_dump[:exit_code].zero?
+          Rails.logger.error(
+            "Failed to create OpenStack database dump: #{db_dump[:stdout_and_stderr]}"
+          )
+          FileUtils.rm_f(dump_path)
+          raise ::Crowbar::Error::UpgradeDatabaseDumpError.new(
+            db_dump[:stdout_and_stderr]
+          )
+        end
+
+        ::Crowbar::UpgradeStatus.new.end_step
+      rescue ::Crowbar::Error::UpgradeNotEnoughDiskSpaceError => e
+        ::Crowbar::UpgradeStatus.new.end_step(
+          false,
+          backup_openstack: {
+            data: e.message,
+            help: "Make sure you have enough disk space to store the OpenStack database dump."
+          }
+        )
+      rescue ::Crowbar::Error::UpgradeFreeDiskSpaceError,
+             ::Crowbar::Error::UpgradeDatabaseSizeError,
+             ::Crowbar::Error::UpgradeDatabaseDumpError => e
+        ::Crowbar::UpgradeStatus.new.end_step(
+          false,
+          backup_openstack: {
+            data: e.message,
+            help: "Please refer to the error message in the response."
```

Comment:
I'm wondering if, in the case of such generic "help texts", we shouldn't rather omit passing "help" at all. AFAIK from the UI point the "help" key is optional, right @skazi0 ?